semantics/src/traversal/expressions.rs

```diff
@@ -296,97 +289,157 @@ fn expr_call(
-    argument_attributes: Vec<ExpressionAttributes>,
+    context: Shared<Context>,
+    typ: Type,
+    args: &Spanned<Vec<Spanned<fe::CallArg>>>,
 ) -> Result<ExpressionAttributes, SemanticError> {
-    let contract_scope = &scope.borrow().contract_scope();
-    let called_func = contract_scope
+    // TODO: check that there is only one param
+    call_arg(scope, context, &args.node[0])?;
+    Ok(ExpressionAttributes::new(typ, Location::Value))
+}
+
+fn expr_call_self_attribute(
+    scope: Shared<BlockScope>,
+    context: Shared<Context>,
+    func_name: String,
+    args: &Spanned<Vec<Spanned<fe::CallArg>>>,
+) -> Result<ExpressionAttributes, SemanticError> {
+    if let Some(ContractFunctionDef {
+        is_public: _,
+        param_types,
+        return_type,
+        scope: _,
+    }) = scope
         .borrow()
-        .function_def(func_name.node.to_string());
-
-    match called_func {
-        Some(ContractFunctionDef {
-            is_public: _,
-            param_types,
-            return_type,
-            scope: _,
-        }) => {
-            if fixed_sizes_to_types(param_types)
-                != expression_attributes_to_types(argument_attributes)
-            {
-                return Err(SemanticError::type_error());
-            }
+        .contract_scope()
+        .borrow()
+        .function_def(func_name)
+    {
+        let argument_attributes = args
+            .node
+            .iter()
+            .map(|arg| call_arg(Rc::clone(&scope), Rc::clone(&context), arg))
+            .collect::<Result<Vec<_>, _>>()?;
 
-            Ok(ExpressionAttributes::new(
-                return_type.into(),
-                Location::Value,
-            ))
+        if fixed_sizes_to_types(param_types) != expression_attributes_to_types(argument_attributes)
+        {
+            return Err(SemanticError::type_error());
         }
-        None => Err(SemanticError::undefined_value()),
+
+        let return_location = match &return_type {
+            FixedSize::Base(_) => Location::Value,
+            _ => Location::Memory,
+        };
+        Ok(ExpressionAttributes::new(
+            return_type.into(),
+            return_location,
+        ))
+    } else {
+        Err(SemanticError::undefined_value())
+    }
+}
+
+fn expr_call_value_attribute(
+    scope: Shared<BlockScope>,
+    context: Shared<Context>,
+    func: &Spanned<fe::Expr>,
+    args: &Spanned<Vec<Spanned<fe::CallArg>>>,
+) -> Result<ExpressionAttributes, SemanticError> {
+    if let fe::Expr::Attribute { value, attr } = &func.node {
+        let value_attributes = expr(scope, context, value)?;
+
+        // for now all of these function expect 0 arguments
+        if !args.node.is_empty() {
+            return Err(SemanticError::wrong_number_of_params());
+        }
+
+        return match attr.node {
+            builtins::CLONE => value_attributes.into_cloned(),
+            builtins::TO_MEM => value_attributes.into_cloned_from_sto(),
```

Comment:
So, `.clone()` would be used to do a memory to memory copy and `to_mem()` to make a storage to memory copy? We might have to fine tune the naming.

Reply:
What is a good alternative?

Reply:
> So, .clone() would be used to do a memory to memory copy and to_mem() to make a storage to memory copy? 

And yeah, this is the case.

Reply:
One alternative to consider would be to treat the difference as an implementation detail and just use `clone` for both. If I understand correctly, there isn't really a case where one has to choose between `clone()` or `to_mem()` because each does only apply in one specific scenario where the other one doesn't apply. But we can argue that eventually both methods allocate a new segment of memory. It is just the origin that is different. So maybe we could just branch into two different implementations internally but only expose `clone()`?